### PR TITLE
Faster coverage file matching

### PR DESCRIPTION
**Pull request by kageiit**
- Do not rely on workspace scanning to determine line coverage to path information
- Use diff changed files to obtain relative path from project root during coverage computation
- Refactor code and cleanup a bit
- Delete dead code

### CODE REVIEW

**Comment by kageiit** [comment]
cc @ChaitanyaPramod

File: src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java [hunk 1]
```diff
@@ -79,26 +76,26 @@ protected CodeCoverageMetrics getCoverageMetrics() {
...
             ExecutionFileLoader executionFileLoader = jacocoAction.getJacocoReport().parse(includes, excludes);
             for (IPackageCoverage packageCoverage : executionFileLoader.getBundleCoverage().getPackages()) {
-                for (ISourceFileCoverage fileCoverage : packageCoverage.getSourceFiles()) {
-                    String relativePathFromProjectRoot = getRelativePathFromProjectRoot(pathResolver, fileCoverage);
+                for (IClassCoverage classCoverage : packageCoverage.getClasses()) {
+                    String relativePathFromProjectRoot = getRelativePathFromProjectRoot(
+                            classCoverage.getSourceFileName());
                     if (relativePathFromProjectRoot != null) {
-                        lineCoverage.put(relativePathFromProjectRoot, getPerLineCoverage(fileCoverage));
+                        lineCoverage.put(relativePathFromProjectRoot, getPerLineCoverage(classCoverage));
```
Comment (artms) [inline src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java]: Won't it change result behaviour? Looks like you're changing fileCoverage to class coverage? Is this related to performance?

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java [hunk 2]
```diff
@@ -70,75 +73,61 @@ public InputSource resolveEntity(String publicId, String systemId) throws SAXExc
...
+                String finalFileName = CoverageProvider.getRelativePathFromProjectRoot(includeFiles, fileName);
+                if (finalFileName != null) {
+                    fileNames.add(finalFileName);
+                    childNodes.add(classNode.getChildNodes());
                 }
-                fileNames.add(fileName);
-                childNodes.add(classNode.getChildNodes());
             }
 
-            // Make multiple guesses on which of the `sourceDirs` contains files in question
```
Comment (artms) [inline src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java]: Is this logic still in place?

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java [hunk 3]
```diff
@@ -20,61 +20,32 @@
...
 
-    public void setWorkspace(FilePath workspace) {
-        this.workspace = workspace;
-    }
+public abstract class CoverageProvider {
 
-    protected FilePath getWorkspace() {
-        return workspace;
-    }
+    private static final String DEFAULT_COVERAGE_REPORT_PATTERN = "**/coverage*.xml, **/cobertura*.xml";
```
[the remark this answers is not in the document]
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java]: I plan to add a jacoco parser to not require the jacoco plugin. Which is why I left it here

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java [hunk 4]
```diff
@@ -85,12 +56,33 @@ String getCoverageReportPattern() {
...
+        return getRelativePathFromProjectRoot(includeFiles, file);
+    }
+
+    @Nullable
+    static String getRelativePathFromProjectRoot(Map<String, String> includeFiles, String file) {
+        String finalFile = null;
+        if (includeFiles == null || includeFiles.isEmpty()) {
+            finalFile = file;
+        } else {
+            String includedFileName = FilenameUtils.getName(file);
```
[the remark this answers is not in the document]
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java]: That is the same case even before with cobertura and they way it used to guess the filename based on the package name + class name (what jacoco also does today). The issue is with languages like kotlin/Scala which can have multiple top level classes in a file, they will not be able to get code coverage since package path + source file name will not match where they are actually present in the filesystem. 

A better approach might be to do two separate marches i.e one with just file name and another with package name + sourcefile name. This way multiple classes with same name and different packages will match correctly. In the case the full match is not possible, we fall back to South file name match and pick just one. The fact that we only check against files part of a diff means that the possibility of a bad match is very unlikely (only if two files with same name are touched as part of the diff), but that will the beat we can do.

File: src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java [hunk 5]
```diff
@@ -79,26 +76,26 @@ protected CodeCoverageMetrics getCoverageMetrics() {
...
-            excludes = new String[]{jacocoPublisher.getExclusionPattern()};
+            excludes = new String[] {jacocoPublisher.getExclusionPattern()};
         }
 
         try {
             ExecutionFileLoader executionFileLoader = jacocoAction.getJacocoReport().parse(includes, excludes);
             for (IPackageCoverage packageCoverage : executionFileLoader.getBundleCoverage().getPackages()) {
-                for (ISourceFileCoverage fileCoverage : packageCoverage.getSourceFiles()) {
-                    String relativePathFromProjectRoot = getRelativePathFromProjectRoot(pathResolver, fileCoverage);
+                for (IClassCoverage classCoverage : packageCoverage.getClasses()) {
```
[the remark this answers is not in the document]
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java]: I think this change can be removed if isourcefilenode can give us the correct source file name for marching against include files

File: src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java [hunk 6]
```diff
@@ -79,26 +76,26 @@ protected CodeCoverageMetrics getCoverageMetrics() {
...
+                    String relativePathFromProjectRoot = getRelativePathFromProjectRoot(
```
[the remark this answers is not in the document]
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java]: See response in a previous comment above

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java (hunk 2 above)
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java]: We don't need this logic anymore

File: src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java (hunk 1 above)
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/JacocoCoverageProvider.java]: Going to revert the behavior back

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java [hunk 7]
```diff
@@ -70,75 +73,61 @@ public InputSource resolveEntity(String publicId, String systemId) throws SAXExc
...
 
-        return parse(coverageData);
+        return parse(coverageData, includeFiles);
     }
 
-    private Map<String, List<Integer>> parse(Map<NodeList, List<String>> coverageData) {
+    private static Map<String, List<Integer>> parse(Set<NodeList> coverageData, Map<String, String> includeFiles) {
         Map<String, SortedMap<Integer, Integer>> internalCounts = new HashMap<String, SortedMap<Integer, Integer>>();
 
         // Each entry in the map is an XML list of classes (files) mapped to its possible source roots
```
[the remark this answers is not in the document]
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParser.java]: Good Catch

File: src/test/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParserTest.java [hunk 8]
```diff
@@ -117,6 +106,20 @@ public void testGetLineCoverageWithIncludes()
 
     }
 
+    @Test
+    public void testGetLineCoverageWithMultipleIncludes()
```
Comment (kageiit) [inline src/test/java/com/uber/jenkins/phabricator/coverage/CoberturaXMLParserTest.java]: @ChaitanyaPramod This test checks the case you mentioned with multiple files having same name but different package. I only added it for cobertura as the test resource was easy to obtain. Feel free to add a test after this merges for the jacoco use case

I will be adding a test for the jacoco file provider i.e the one which does not rely on jekins jacoco plugin in a follow up change

**Comment by kageiit** [comment]
@ChaitanyaPramod Addressed your concerns and added new tests. PTAL

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java [hunk 9]
```diff
@@ -83,14 +55,70 @@ String getCoverageReportPattern() {
...
+            String maxMatchFile = coverageFile;
+            for (String includedFile : includeFiles) {
+                int currmatch = suffixMatch(includedFile, coverageFile);
+                if (currmatch > maxMatch) {
+                    maxMatch = currmatch;
+                    maxMatchFile = includedFile;
+                }
+            }
+
+            // Make sure the match is atleast a full match on the filename and not an accidental partial match
```
[the remark this answers is not in the document]
Comment (kageiit) [reply src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java]: Good point. I will make the change

File: src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java [hunk 10]
```diff
@@ -83,14 +55,70 @@ String getCoverageReportPattern() {
...
+            if (FilenameUtils.getName(maxMatchFile).equals(FilenameUtils.getName(coverageFile))) {
+                return maxMatchFile;
+            } else {
+                return coverageFile;
+            }
+        }
+    }
+
+    private static int suffixMatch(String changedFile, String coverageFile) {
```
Comment (ChaitanyaPramod) [inline src/main/java/com/uber/jenkins/phabricator/coverage/CoverageProvider.java]: Just noticed that
```java
suffixMatch("pathA/file.ext", "pathB/file.ext") == 9
```

Perhaps we could simplify to
```java
int prefixLength = changedFile.endsWith(coverageFile) ? changedFile.lastIndexOf(coverageFile) : -1
// -1 signals no match
```
